### ISSUE TITLE
Use-after-free of MPI communicators

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#2533]
This is from a post by Daniel Arndt on the mailing list https://groups.google.com/forum/#!topic/dealii/SyFQDV_eh1E. I'm moving this here to be able to track progress on the issue.

The following exception has been observed:
```
*** An error occurred in MPI_Allreduce
*** on communicator MPI_COMM_WORLD
*** MPI_ERR_COMM: invalid communicator
```
with the call stack
```
#0  0x00000000006d4608 in ?? ()
#1  0x00007fffee4b2775 in PMPI_Allreduce () from /usr/lib/libmpi.so.1
#2  0x00007fffeed72db1 in Epetra_MpiComm::MinAll (this=<optimised out>, PartialMins=<optimised out>, GlobalMins=<optimised out>, Count=<optimised out>) at /mnt/data/darndt/Sources/trilinos-12.6.1-Source/packages/epetra/src/Epetra_MpiComm.cpp:189
#3  0x00007fffeecb11c2 in Epetra_BlockMap::SameAs (this=0x6d3e00, Map=...) at /mnt/data/darndt/Sources/trilinos-12.6.1-Source/packages/epetra/src/Epetra_BlockMap.cpp:772
#4  0x00007ffff510ed04 in dealii::TrilinosWrappers::MPI::Vector::reinit (this=0x88ad20, v=..., omit_zeroing_entries=omit_zeroing_entries@entry=true, allow_different_maps=allow_different_maps@entry=false) at ../source/lac/trilinos_vector.cc:247
#5  0x00000000004ccecf in dealii::MGSmootherPrecondition<dealii::TrilinosWrappers::SparseMatrix, dealii::PreconditionIdentity, dealii::TrilinosWrappers::MPI::Vector>::smooth (this=this@entry=0x7fffffffca10, level=level@entry=1, u=..., rhs=...)
    at /mnt/data/darndt/deal.II-periodic/include/deal.II/multigrid/mg_smoother.h:985
#6  0x00007ffff5b61500 in dealii::Multigrid<dealii::TrilinosWrappers::MPI::Vector>::level_v_step (this=this@entry=0x7fffffffcb90, level=1) at ../include/deal.II/multigrid/multigrid.templates.h:167
#7  0x00007ffff5b7fe3c in dealii::Multigrid<dealii::TrilinosWrappers::MPI::Vector>::cycle (this=this@entry=0x7fffffffcb90) at ../include/deal.II/multigrid/multigrid.templates.h:415
#8  0x00000000004cbbcb in MyLaplace<2>::solve (this=this@entry=0x7fffffffd0e0) at source/LaplaceMeshworker.cc:186
#9  0x00000000004cc020 in MyLaplace<2>::run (this=this@entry=0x7fffffffd0e0) at source/LaplaceMeshworker.cc:207
#10 0x00000000004af689 in main (argc=1, argv=0x7fffffffe0a8) at source/LaplaceMeshworker.cc:226
```

A look with valgrind reveals a use-after-free problem:
```
Invalid read of size 1
   at 0x229FFDDC: PMPI_Allreduce (in /usr/lib64/openmpi/lib/libmpi.so.1.6.2)
   by 0x1CC95B1D: Epetra_MpiComm::MinAll(int*, int*, int) const (in /home/kronbichler/sw/lib/libepetra.so.12.3)
   by 0x1CBBC403: Epetra_BlockMap::SameAs(Epetra_BlockMap const&) const (in /home/kronbichler/sw/lib/libepetra.so.12.3)
   by 0xAFB0563: dealii::TrilinosWrappers::MPI::Vector::reinit(dealii::TrilinosWrappers::VectorBase const&, bool, bool) (trilinos_vector.cc:247)
   by 0x45AF0E: dealii::MGSmootherPrecondition<dealii::TrilinosWrappers::SparseMatrix, dealii::PreconditionIdentity, dealii::TrilinosWrappers::MPI::Vector>::smooth(unsigned int, dealii::TrilinosWrappers::MPI::Vector&, dealii::TrilinosWrappers::MPI::Vector const&) const (mg_smoother.h:985)
   by 0xBA90B2B: dealii::Multigrid<dealii::TrilinosWrappers::MPI::Vector>::level_v_step(unsigned int) (multigrid.templates.h:149)
   by 0xBAAD215: dealii::Multigrid<dealii::TrilinosWrappers::MPI::Vector>::cycle() (multigrid.templates.h:397)
   by 0x45BDD2: MyLaplace<2>::solve() (LaplaceMeshworker.cc:186)
   by 0x45C202: MyLaplace<2>::run() (LaplaceMeshworker.cc:207)
   by 0x439DF3: main (LaplaceMeshworker.cc:223)
 Address 0x2aa7ad48 is 152 bytes inside a block of size 808 free'd
   at 0x4C29D6A: free (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x229DBA8B: ompi_comm_free (in /usr/lib64/openmpi/lib/libmpi.so.1.6.2)
   by 0x22A06ECE: PMPI_Comm_free (in /usr/lib64/openmpi/lib/libmpi.so.1.6.2)
   by 0xBB215F2: dealii::parallel::Triangulation<2, 2>::~Triangulation() (tria_base.cc:94)
   by 0xBB18C92: dealii::parallel::distributed::Triangulation<2, 2>::~Triangulation() (tria.cc:2241)
   by 0x43A02B: ~MyLaplace (LaplaceMeshworker.cc:43)
   by 0x43A02B: main (LaplaceMeshworker.cc:222)
```

The funny thing was that I could reproduce the problem on a system with OpenMPI but not on another one with MPICH, which suggests that the allocation is done inside MPI and implementation-dependent. On a higher level, the reason for the crash is that the call to TrilinosWrappers::MPI::Vector::reinit checks whether the underlying map of the incoming vector (call in MGSmootherPrecondition) is the same as what was stored in the Trilinos vector before. What happens is that the MPI communicator in the 'old' incarnation of the vector in the memory pool refers the MPI_Comm object from parallel::distributed::Triangulation (cloned from MPI_COMM_WORLD), but that old MPI_Comm object had been free upon destruction of parallel::Triangulation (source/distributed/tria_base.cc:94).

To solve the problem, we need to avoid a dangling reference to an MPI_Comm that gets destroyed. But it is tricky:
- The cause for the problem is that the parallel Triangulation frees the communicator that is in use elsewhere. Ideally, we would like to have all communicators wrapped into a Subscriptor that would then raise an exception when the free happens. However, I consider this overkill because would need to go through all our classes (and external interface) to track these. Then, I am still not sure whether we would find them all because we often go over to external libraries.
- We might need to call Utilities::MPI::duplicate_communicator more often to avoid the dangling reference. However, that means that we to properly destruct those communicators again all over the place. This requires writing a wrapper to MPI_Comm and introduce this in many places. Again, lots of work. The advantage of this would be that we avoided the problem of the dangling reference (rather than just get an exception as above). A disadvantage would be that I'm not sure whether all algorithms really support different communicators: Think about a matrix-vector product where there can be three different communicators. Of course, they would all be congruent so the pattern should go through, but are we sure all packages support this? I personally don't know if even our parallel vectors with a quite limited MPI part work for that case, not to mention Trilinos with all those different places. There would be different Epetra_MpiComm objects all over the place. In case there would be inconsistencies, we would need to put lots of MPI_Comm_compare calls to catch those before going into external libraries. A nightmare.

One might argue that a properly written code frees all vectors before starting with a new Triangulation (we might even make that a requirement and not put any guarantees). But that brings us to the next problem: A global pool for vectors as GrowingVectorMemory is problematic because it carries over state in an unexpected way. I don't particularly like that. It would be much clearer if classes that need to keep vectors around for performance reasons would hold them explicitly. Be it iterative solvers or the MGSmootherPrecondition. Ideally, the destructor of the triangulation could anticipate that there might be vectors depending on the communicator and hence clear the GrowingVectorMemory object. But it would be necessary to do this for all vector types currently available. I don't know if that is feasible.

So there is nothing related to MGSmootherPrecondition that causes this problem. It is rather GrowingVectorMemory which is unsafe. It could happen with all vectors that carry an MPI communicator that does not get properly duplicated, including our own parallel::distributed::Vector. Could we make the `pool` variable in `GrowingVectorMemory` non-static? (I know it would lead to more vectors to be re-allocated, but I prefer to let the user manage those things. A positive aspect of this would be that we could avoid putting all those extra de-allocations in the destructor of MPI_InitFinalize.)

Any opinions what approach we should take?

**Comment by tjhei** [comment]
> Think about a matrix-vector product where there can be three different communicators. Of course, they would all be congruent so the pattern should go through, but are we sure all packages support this?

I vaguely remember that I had some trouble with our PETSc-wrappers because of this. I would very much prefer to have the same communicator when doing an operation that involves more than one object.

>  It is rather GrowingVectorMemory which is unsafe.

I don't particularly like this class and the way it behaves. Creating vectors can't be that expensive, right? The only point where it makes sense to reuse vectors is within a single linear solve.

Question: if we make the change to GrowingVectorMemory, would we still run into the same problem in MGSmoother but only if somebody decides to keep the MGSmoother object alive between destruction of Triangulation, right?

One more thing: would it help to check if the communicator is identical before calling Epetra_BlockMap::SameAs?

**Comment by kronbichler** [comment]
> I don't particularly like this class [GrowingVectorMemory] and the way it behaves. Creating vectors can't be that expensive, right? 

It's usually not. But the case of MGSmootherPreconditioner is actually one of those cases where you would like to retain the vectors over several calls and where memory allocation or unnecessary initialization will show up, so I do see the point of @guidokanschat for introducing something here. But I would propose to add a class variable for that (or we transition to the solution proposed in #2534).

> Question: if we make the change to GrowingVectorMemory, would we still run into the same problem in MGSmoother but only if somebody decides to keep the MGSmoother object alive between destruction of Triangulation, right?

Correct. If somebody keeps alive any data structure that holds the MPI_Comm from the triangulation that is not duplicated, e.g. sparse matrices, vectors, or smoothers, the issue can potentially be the same. Depending on how the MPI library implements `MPI_Comm_dup`, it can be a silent problem or fail miserably. If we go down the route to changing GrowingVectorMemory and cleaning some things up but otherwise not change the communicator handling, we have to put bold warnings about the fact that whenever a triangulation goes out of scope, all related objects that used the communicator of the that triangulation should also go out of scope.

We could at least partially help here by subscribing the objects that the library creates to the parallel triangulation. (The user could still mess things up by using `triangulation.get_communicator()` but Daniel's code should actually be fine to keep vectors beyond the lifetime of the triangulation because he uses MPI_COMM_WORLD in all of his code.)

> One more thing: would it help to check if the communicator is identical before calling Epetra_BlockMap::SameAs?

No, we would access invalid memory in the `MPI_Comm_compare` call that we would need at this place.

**Comment by tjhei** [comment]
Reusing temporary vectors within a linear solve (and all contained solves/preconditioner application/MG) makes a lot of sense of course. My point was that it is unlikely to be worth attempting reuse outside of that. There is no chance that reuse can happen between mesh changes anyways for example. So maybe the lifetime of a VectorMemory should be smaller than the lifetime of the triangulation, needs to be as big as a linear solve, and should typically be as long as the mesh doesn't change.

I like the idea of making the vector memory more explicit. What would be a good default though if the user doesn't want to care? Or should we force the user to supply a VectorMemory for preconditioners and all MG classes?

> We could at least partially help here by subscribing the objects that the library creates to the parallel triangulation.

Can you give examples for objects that would fall under this that currently don't do this?

> No, we would access invalid memory in the MPI_Comm_compare call that we would need at this place.

I was thinking of comparing the two comms by equality directly.

**Comment by kronbichler** [comment]
> What would be a good default though if the user doesn't want to care? Or should we force the user to supply a VectorMemory for preconditioners and all MG classes?

My favorite approach is to put the vectors we use in MG classes as member variables. For the coarse solver, we force the user to provide memory already anyway, so if one cares about speed one would provide GrowingVectorMemory at the constructor of the solver already so that it gets reused in every linear iteration. In general, in algorithms such as GMRES you really want to have a memory class, but we are quite explicit there already so changing the behavior in the GrowingVectorMemory class should be enough. If a user cares about performance, one can always keep the GrowingVectorMemory around for several calls.

> > We could at least partially help here by subscribing the objects that the library creates to the parallel triangulation.

> Can you give examples for objects that would fall under this that currently don't do this?

Potentially it's all places inside the library that initialize a vector using a communicator that comes from the triangulation. But we can only protect those parts that hold the vector structure after that. I few things in multigrid come to mind. Apart from that, probably not very much. At least I couldn't grep any offender outside multigrid (even though MatrixFree should probably be attached, too).

> I was thinking of comparing the two comms by equality directly.

That's of course possible :+1:

**Comment by bangerth** [comment]
Can you explain the problem to me again in different words? Was the case that `p::d::Triangulation` duplicated a communicator, then we created some vectors with this duplicated communicator, then the triangulation goes out of scope and destroys the communicator, and the communicator used by the vector becomes stale? Yes, that would be a problem.

Many years ago, I wrote the `duplicate_communicator()` function and used it everywhere to give every object its own, unique communicator so that (i) I avoid these sort of lifetime problems, (ii) I could better detect problems of multiple parts of programs sending communications from multiple threads, as they could collide if sent on the same communicator. But I had to abandon this approach and undo all of the changes when, after several weeks of extremely frustrating debugging, I came to realize that my application hung once I had created 64k communicators -- hours and thousands of time steps into a run :-(

**Comment by bangerth** [comment]
For the record, I would not be opposed to a solution whereby `GrowingVectorMemory::free` actually destroys a vector if the vector class declares that it carries references to other objects (e.g., via a static member variable). The MPI vectors would do this, whereas all other vectors do not as they are neatly self-contained.

**Comment by Rombur** [comment]
I am going to ask a stupid question but why can't we use a shared pointer for the communicator? The triangulation could go out of scope but the communicator should stay alive as long as someone is using it.

**Comment by tjhei** [comment]
> For the record, I would not be opposed to a solution whereby GrowingVectorMemory::free actually destroys a vector if the vector class declares that it carries references to other objects

But then there is no advantage of using GrowingVectorMemory and you would do a ton of expensive vector creations inside solvers like GMRES (and GMG, ...). I don't think this is feasible.

> I am going to ask a stupid question but why can't we use a shared pointer for the communicator? The triangulation could go out of scope but the communicator should stay alive as long as someone is using it.

All constructors take a const ref to a communicator right now, so your suggest would require changing the interface of every class.

**Comment by bangerth** [comment]
On 04/21/2016 07:49 AM, Timo Heister wrote:
>     For the record, I would not be opposed to a solution whereby
>     GrowingVectorMemory::free actually destroys a vector if the vector class
>     declares that it carries references to other objects
>
> But then there is no advantage of using GrowingVectorMemory and you would do a
> ton of expensive vector creations inside solvers like GMRES (and GMG, ...). I
> don't think this is feasible.
Not within one call to GMRES if we structure it right. But you would indeed 
lose *across* subsequent calls to GMRES.

**Comment by kronbichler** [comment]
> > > For the record, I would not be opposed to a solution whereby GrowingVectorMemory::free actually destroys a vector if the vector class declares that it carries references to other objects

> > But then there is no advantage of using GrowingVectorMemory and you would do a ton of expensive vector creations inside solvers like GMRES (and GMG, ...). I don't think this is feasible.

> Not within one call to GMRES if we structure it right. But you would indeed lose *across* subsequent calls to GMRES.

I don't understand the problem here: If we make the pool of vectors in GrowingVectorMemory a usual member variable rather than static, then a user can decide to keep a variable of type GrowingVectorMemory in case he feels for it. But I wouldn't worry about a user keeping GrowingVectorMemory beyond the destruction of a distributed triangulation - that's the same way as keeping a vector beyond destruction of triangulation and we can document that this will not work. It will work if a user keeps his own MPI_Comm but not if he uses the data structures from the triangulation. We as a library do the right thing of having our copy. Unfortunately we cannot change the interface to MPI_Comm now any more to have more expressive assertions of subscriptor-type. But I don't think it's that common either, so a clear documentation and the immediate fix in GrowingVectorMemory would do the trick in my opinion.

**Comment by bangerth** [comment]
But all of the `GrowingVectorMemory` objects draw from the same memory pool (which is a static member variable). Letting one of the `GrowingVectorMemory` objects go out of scope does not release the allocated vectors.

**Comment by bangerth** [comment]
Oh, or maybe I misunderstood -- you want to break this, by making the pool a regular member variable? That would of course work, but then you won't get any benefit across calls to functions that internally use this class. That would negate most of the purpose of the class.

If that's what you want, then I'd much rather go with my idea of releasing memory from the pool only for vectors that have state. Or maybe a middle ground, where each `GrowingVectorMemory` object pulls from a global pool, but keeps track which of these vectors it got out of the pool and releases those it allocated if the vectors have state.

**Comment by kronbichler** [comment]
Indeed, I was thinking about breaking the global pool variable and put things into a regular one. The middle ground solution would probably be the right way to go. I'll think how this can be done.

**Comment by kronbichler** [comment]
I've started drafting a new member `VectorType::reinit_uses_nontrivial_state`. While we can introduce those methods to all of our vectors, I am a bit hesitant to also force users who might define their own vector class (and use it with GrowingVectorMemory). We also have another possibility: We either build some SFINAE magic similar to the IsBlockVector type in vector_memory.templates.h that can handle the case of a vector that does not have that magic member variable (for which we can be either conservative and free vectors or aggressive). What do you prefer?

**Comment by bangerth** [comment]
I'd say just go ahead and require the new field. The number of people who write their own vectors is probably small, and there is a backward compatible solution: just add the field, it is simply unused with older versions of deal.II.

**Comment by tjhei** [comment]
Wouldn't it be enough to check the communicator inside the vector reinit() for equality? Then we don't need any changes inside GrowingVectorMemory.

**Comment by bangerth** [comment]
> Wouldn't it be enough to check the communicator inside the vector reinit() for
> equality?

The issue was that the communicator stored in the vector may no longer exist. 
(The communicator is in essence a pointer that may point to a region of memory 
that is no longer valid.)

**Comment by kronbichler** [comment]
> Wouldn't it be enough to check the communicator inside the vector reinit() for equality? Then we don't need any changes inside GrowingVectorMemory.

I agree with @tjhei that we should try this. I will set up a test that gets through the various possibilities and check if I can get it run without invalid memory accesses. I would of course prefer if we manage to do it without the extra static variable.

**Comment by kronbichler** [comment]
As seen in #2548, it seems that I do not need further fixes than make the vector reinit more conservative. Thus, I prefer this solution over the change in GrowingVectorMemory. (Thus, my initial fear turned out to be premature and it is pretty simple to fix anyway.)
